bot/exts/info/doc/_batch_parser.py

```diff
@@ -0,0 +1,206 @@
+                        # Some items are present in the inventories multiple times under different symbol names,
+                        # if we already parsed an equal item, we can just skip it.
+                        continue
+
+                    markdown = await bot.instance.loop.run_in_executor(None, get_symbol_markdown, soup, item)
+                    if markdown is not None:
+                        scheduling.create_task(doc_cache.set(item, markdown))
+                    else:
+                        scheduling.create_task(self.stale_inventory_notifier.send_warning(item))
+                except Exception as e:
+                    log.exception(f"Unexpected error when handling {item}")
+                    future.set_exception(e)
+                else:
+                    future.set_result(markdown)
+                await asyncio.sleep(0.1)
+        finally:
+            self._parse_task = None
+            log.trace("Finished parsing queue.")
+
+    def _move_to_front(self, item: Union[QueueItem, _cog.DocItem]) -> None:
+        """Move `item` to the front of the parse queue."""
+        # The parse queue stores soups along with the doc symbols in QueueItem objects,
+        # in case we're moving a DocItem we have to get the associated QueueItem first and then move it.
+        item_index = self._queue.index(item)
+        queue_item = self._queue.pop(item_index)
+
+        self._queue.append(queue_item)
+        log.trace(f"Moved {item} to the front of the queue.")
+
+    def add_item(self, doc_item: _cog.DocItem) -> None:
+        """Map a DocItem to its page so that the symbol will be parsed once the page is requested."""
+        self._page_doc_items[doc_item.url].append(doc_item)
+
+    async def clear(self) -> None:
+        """
+        Clear all internal symbol data.
+
+        All currently requested items are waited to be parsed before clearing.
+        """
+        for future in filter(attrgetter("user_requested"), self._item_futures.values()):
+            await future
+        if self._parse_task is not None:
+            self._parse_task.cancel()
+        self._queue.clear()
+        self._page_doc_items.clear()
+        self._item_futures.clear()
+
+    async def _clean_up_futures(self) -> None:
+        """
+        Clear old futures from internal results.
+
+        After a future is set, we only need to wait for old requests to its associated `DocItem` to finish
+        as all new requests will get the value from the redis cache in the cog first.
+        Keeping them around for longer than a second is unnecessary and keeps the parsed Markdown strings alive.
+        """
+        while True:
+            if not self._queue:
+                current_time = time.time()
+                for key, future in self._item_futures.copy().items():
+                    if current_time - future.result_set_time > 5:
+                        del self._item_futures[key]
+            await asyncio.sleep(5)
```

Comment:
I understand this is necessary, but it's quite ugly. Why don't we just have the queue consumer delete all futures for which `user_requested == False`? We know when it's `False`, no one is waiting for the future. This is a lot simpler. The downside is that if someone requests a symbol soon after it is parsed, they'll still have to take the route of getting it through redis. Technically slower, but not a big impact IMO. I don't image redis is noticeably slower.

Reply:
I've removed it in fbe52273 and replaced it with a simpler system. A future is created either when an user requests it or the parsing of its associated doc item begins, then it is kept in the dict until its result is set.
I've replaced the redis set task with a normal await to prevent a potential race condition when the cache wouldn't be set before the item deletion and an another request for the item came in from an user.
I believe it didn't introduce any race conditions or logic errors but would appreciate a more thorough look